File: cli/presentation.py
from __future__ import annotations

from collections.abc import Iterable, Sequence
from pathlib import Path

import codegrinder_pb2 as pb

from helpers import abbreviate_home, course_directory, dashes


def sorted_assignment_items(items: Iterable[pb.AssignmentListItem]) -> list[pb.AssignmentListItem]:
    return sorted(
        items,
        key=lambda item: (
            item.assignment.course_id,
            item.due_at.seconds if item.HasField("due_at") else 0,
            item.lock_at.seconds if item.HasField("lock_at") else 0,
            item.assignment.user_id,
            item.assignment.problem_set_id,
        ),
    )


def sorted_student_assignment_items(items: Iterable[pb.AssignmentListItem]) -> list[pb.AssignmentListItem]:
    return sorted(
        items,
        key=lambda item: (
            item.assignment.user_id,
            item.assignment.course_id,
            item.due_at.seconds if item.HasField("due_at") else 0,
            item.assignment.problem_set_id,
        ),
    )


def _pretty_assignment_path(workspace_root: Path, item: pb.AssignmentListItem) -> str:
    path = workspace_root.expanduser() / course_directory(item.course_name) / item.assignment.problem_set_id
    return abbreviate_home(path)


def _assignment_title(item: pb.AssignmentListItem) -> str:
    if item.assignment_title:
        return item.assignment_title
    if item.problem_set_note:
        return item.problem_set_note
    return item.assignment.problem_set_id


def print_assignment_list(items: Sequence[pb.AssignmentListItem], workspace_root: Path) -> None:
    titles = [_assignment_title(item) for item in items]
    longest_title = max(len(title) for title in titles)

    current_course_id = ""
    for item, title in zip(items, titles, strict=True):
        if item.assignment.course_id != current_course_id:
            if current_course_id != "":
                print()
            current_course_id = item.assignment.course_id
            print(item.course_name)
            print(dashes(len(item.course_name)))

        percent = round(float(item.assignment_score) * 100)
        pretty_path = _pretty_assignment_path(workspace_root, item)
        suffix = ""
        if item.download_status == pb.ASSIGNMENT_DOWNLOAD_STATUS_PREREQ_NOT_READY:
            suffix = f" waiting for {item.prerequisite_problem_set_id}"
        print(f"{title:<{longest_title}}  {percent:>3}% ({pretty_path}){suffix}")


def print_problem_catalog(problem_sets: Sequence[pb.ProblemCatalogSet], host: str) -> None:
    for index, pset in enumerate(problem_sets):
        if index > 0:
            print()
        print(pset.problem_set_note)

        for problem in pset.problems:
            if problem.problem_weight == 1:
                print(f"  * {problem.problem_note} ({problem.problem_id})")
            else:
                print(f"  * {problem.problem_note} ({problem.problem_id}, weight {problem.problem_weight})")
            for step in problem.steps:
                text = step.step_note.replace("\n", "\n       ")
                suffix = "" if step.step_weight == 1 else f" (weight {step.step_weight})"
                n = int(step.step_number)
                print(f"    {n}. {text}{suffix}")

        print()
        print(f"  → https://{host}/lti/problem_sets/cli/{pset.problem_set_id}")

File: grind/src/presentation.rs
use crate::config::{abbreviate_home, home_dir};
use crate::proto::codegrinder::{AssignmentDownloadStatus, AssignmentListItem, ProblemCatalogSet};
use std::path::{Path, PathBuf};

pub fn course_directory(label: &str) -> String {
    let bytes = label.as_bytes();
    let mut letters = String::new();
    let mut index = 0;
    while index < bytes.len() && bytes[index].is_ascii_alphabetic() {
        letters.push(bytes[index] as char);
        index += 1;
    }
    while index < bytes.len() && (bytes[index] == b'-' || bytes[index] == b' ') {
        index += 1;
    }
    let mut digits = String::new();
    while index < bytes.len() && bytes[index].is_ascii_alphanumeric() {
        digits.push(bytes[index] as char);
        index += 1;
    }
    if !letters.is_empty() && !digits.is_empty() {
        format!("{letters}{digits}").to_lowercase()
    } else {
        label.to_string()
    }
}

pub fn sorted_assignment_items(mut items: Vec<AssignmentListItem>) -> Vec<AssignmentListItem> {
    items.sort_by(|left, right| assignment_sort_key(left).cmp(&assignment_sort_key(right)));
    items
}

pub fn sorted_student_assignment_items(
    mut items: Vec<AssignmentListItem>,
) -> Vec<AssignmentListItem> {
    items.sort_by(|left, right| {
        student_assignment_sort_key(left).cmp(&student_assignment_sort_key(right))
    });
    items
}

pub fn print_assignment_list(items: &[AssignmentListItem], workspace_root: &Path) {
    let titles: Vec<String> = items.iter().map(assignment_title).collect();
    let longest_title = titles.iter().map(String::len).max().unwrap_or(0);
    let mut current_course_id = String::new();
    for (item, title) in items.iter().zip(titles.iter()) {
        let assignment = item.assignment.as_ref();
        let course_id = assignment.map_or("", |assignment| assignment.course_id.as_str());
        if course_id != current_course_id {
            if !current_course_id.is_empty() {
                println!();
            }
            current_course_id = course_id.to_owned();
            println!("{}", item.course_name);
            println!("{}", "-".repeat(item.course_name.len()));
        }
        let percent = (item.assignment_score * 100.0).round() as i64;
        let pretty_path = pretty_assignment_path(workspace_root, item);
        let suffix = if item.download_status() == AssignmentDownloadStatus::PrereqNotReady {
            format!(" waiting for {}", item.prerequisite_problem_set_id)
        } else {
            String::new()
        };
        println!("{title:<longest_title$}  {percent:>3}% ({pretty_path}){suffix}");
    }
}

pub fn print_problem_catalog(problem_sets: &[ProblemCatalogSet], host: &str) {
    for (index, pset) in problem_sets.iter().enumerate() {
        if index > 0 {
            println!();
        }
        println!("{}", pset.problem_set_note);
        for problem in &pset.problems {
            if problem.problem_weight == 1 {
                println!("  * {} ({})", problem.problem_note, problem.problem_id);
            } else {
                println!(
                    "  * {} ({}, weight {})",
                    problem.problem_note, problem.problem_id, problem.problem_weight
                );
            }
            for step in &problem.steps {
                let text = step.step_note.replace('\n', "\n       ");
                let suffix = if step.step_weight == 1 {
                    String::new()
                } else {
                    format!(" (weight {})", step.step_weight)
                };
                println!("    {}. {text}{suffix}", step.step_number);
            }
        }
        println!();
        println!("  -> {}/lti/problem_sets/cli/{}", server_base_url(host), pset.problem_set_id);
    }
}

fn server_base_url(host: &str) -> String {
    let host = host.trim_end_matches('/');
    if host.starts_with("http://") || host.starts_with("https://") {
        host.to_string()
    } else {
        format!("https://{host}")
    }
}

pub fn assignment_label(course_name: &str, problem_set_id: &str) -> String {
    format!("{}/{}", course_directory(course_name), problem_set_id)
}

pub fn assignment_directory(root: &Path, course_name: &str, problem_set_id: &str) -> PathBuf {
    root.join(course_directory(course_name)).join(problem_set_id)
}

fn assignment_title(item: &AssignmentListItem) -> String {
    if !item.assignment_title.is_empty() {
        item.assignment_title.clone()
    } else if !item.problem_set_note.is_empty() {
        item.problem_set_note.clone()
    } else {
        item.assignment
            .as_ref()
            .map(|assignment| assignment.problem_set_id.clone())
            .unwrap_or_default()
    }
}

fn pretty_assignment_path(workspace_root: &Path, item: &AssignmentListItem) -> String {
    let assignment = item.assignment.as_ref();
    let root = if workspace_root.as_os_str().is_empty() {
        home_dir()
    } else {
        workspace_root.to_path_buf()
    };
    abbreviate_home(
        &root
            .join(course_directory(&item.course_name))
            .join(assignment.map_or("", |assignment| assignment.problem_set_id.as_str())),
    )
}

fn assignment_sort_key(item: &AssignmentListItem) -> (&str, i64, i64, &str, &str) {
    let assignment = item.assignment.as_ref();
    (
        assignment.map_or("", |assignment| assignment.course_id.as_str()),
        item.due_at.as_ref().map_or(0, |ts| ts.seconds),
        item.lock_at.as_ref().map_or(0, |ts| ts.seconds),
        assignment.map_or("", |assignment| assignment.user_id.as_str()),
        assignment.map_or("", |assignment| assignment.problem_set_id.as_str()),
    )
}

fn student_assignment_sort_key(item: &AssignmentListItem) -> (&str, &str, i64, &str) {
    let assignment = item.assignment.as_ref();
    (
        assignment.map_or("", |assignment| assignment.user_id.as_str()),
        assignment.map_or("", |assignment| assignment.course_id.as_str()),
        item.due_at.as_ref().map_or(0, |ts| ts.seconds),
        assignment.map_or("", |assignment| assignment.problem_set_id.as_str()),
    )
}

#[cfg(test)]
mod tests {
    use super::{course_directory, print_assignment_list};
    use crate::proto::codegrinder::{AssignmentKey, AssignmentListItem};
    use std::path::Path;

    #[test]
    fn course_directory_normalizes_common_course_names() {
        assert_eq!(course_directory("CS-2810 Fall 2026"), "cs2810");
        assert_eq!(course_directory("CS 101"), "cs101");
        assert_eq!(course_directory("CS3520A Section 1"), "cs3520a");
    }

    #[test]
    fn assignment_list_formatting_does_not_panic() {
        let items = vec![AssignmentListItem {
            assignment: Some(AssignmentKey {
                user_id: "u1".to_string(),
                course_id: "c1".to_string(),
                problem_set_id: "sorting".to_string(),
            }),
            course_name: "CS 2810".to_string(),
            assignment_title: "Sorting lab".to_string(),
            assignment_score: 1.0,
            ..AssignmentListItem::default()
        }];
        print_assignment_list(&items, Path::new("/tmp"));
    }
}
